fs/fs_zenfs.cc

```diff
@@ -800,6 +841,38 @@ IOStatus ZenFS::RenameFile(const std::string& source_path,
   return s;
 }
 
+IOStatus ZenFS::LinkFile(const std::string& src, const std::string& dst,
+                         const IOOptions& options, IODebugContext* dbg) {
+  std::shared_ptr<ZoneFile> src_file(nullptr);
+  std::shared_ptr<ZoneFile> dst_file(nullptr);
+  IOStatus s;
+
+  Debug(logger_, "LinkFile: %s to %s\n", src.c_str(), dst.c_str());
+
+  files_mtx_.lock();
+  dst_file = GetFileInternal(dst);
+  if (dst_file != nullptr) {
+    files_mtx_.unlock();
+    return IOStatus::InvalidArgument("Failed to create link, target exists");
+  }
+  src_file = GetFileInternal(src);
+  if (src_file != nullptr) {
+    src_file->SetLinkInfo(dst);
+    files_.insert(std::make_pair(dst, src_file));
+    files_mtx_.unlock();
```

Comment:
We have to hold the lock for the entire duration of the method, or we will race on `src_file`. If the metadata sync fails, the error path may not clean  up properly during a race. @MaisenbacherD is fixing this in #160 - [here](https://github.com/westerndigitalcorporation/zenfs/pull/160/files#diff-b12ba0b7fd2feee53dba42938e4031b32be21c71dc948ba759844c4c49be0a93R440-R467).

I would suggest to wait for that patch set to get merged and then rebase on top of those changes to use the non locking sync method. Then use a `lock_guard` that is dropped before `s = target()->LinkFile(ToAuxPath(src), ToAuxPath(dst), options, dbg);` by returning from the `if` branch.

Reply:
I can rebase on top of it, but if that is going to take more time, we can merge this and I will submit another patch on top of #160 , fixing the lock issue. 
@yhr thoughts ?

Reply:
I think it is better to wait than to merge a concurrency bug :shrug:

Reply:
Yeah, I can rebase on Dennis's patch after it is pulled and resubmit.

Reply:
This is now rebased on concurrency patches.

Reply:
I think you need to rebase on master since the concurrency patches are already merged.

Reply:
I will wait for the pkg-config to get merged and then rebase on top of it.